### PR TITLE
REF: implement groupby.ops.WrappedCythonFunc

### PR DESCRIPTION
**Pull request by jbrockmendel**
I think this is much closer to "the correct" abstraction here (i find the layered abstractions with similar names in core.groupby hard to follow).  Contains \approx 0 groupby logic (takes ngroups as an argument in some methods) and is only about how to wrap/call/unwrap the libgroupby functions.

There are a few more parts of _cython_operation I'd like to move, but I'm holding off on that pending some non-refactor work that touches those pieces.

No specific plans, but I'm hopeful that something like this can be re-used to de-duplicate logic in nanops and maybe algorithms.

This uses classmethods/staticmethods (i like the explicit statelessness) and passes how+kind as arguments, but an alternative would be to instantiate with how+kind passed to `__init__`.

The only potentially-logical change is that when defining `out_dtype`, this returns "float64" instead of "float".  I haven't checked if that makes a difference on 32bit builds, but after weeks of dealing with np.intp much prefer the explicitness.

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -637,43 +671,25 @@ def _cython_operation(
...
-        if how == "rank":
-            out_dtype = "float"
-        else:
-            if is_numeric:
-                out_dtype = f"{values.dtype.kind}{values.dtype.itemsize}"
-            else:
-                out_dtype = "object"
 
-        codes, _, _ = self.group_info
+        out_shape = WrappedCythonOp.get_output_shape(how, kind, ngroups, values)
```
Comment (jreback) [inline pandas/core/groupby/ops.py:680]: i would go one step further and use an instance of WrappedCythonOp here, then you can pass whatever you need into the construct and just ask the instance for properties (if you want to make this simple), you can use a dataclass for example.

I find this pattern *better* than using a class like this (meaning the class is just a collection of methods here).

**Review by jreback — CHANGES_REQUESTED** [review]
+1 on this

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: updated + green

File: pandas/core/groupby/ops.py [hunk 2]
```diff
@@ -97,54 +97,172 @@
...
-        elif dtype == object:
-            if "object" not in f.__signatures__:
-                # raise NotImplementedError here rather than TypeError later
+
+class WrappedCythonOp:
+    """
+    Dispatch logic for functions defined in _libs.groupby
+    """
+
+    def __init__(self, kind: str, how: str):
```
Comment (jreback) [inline pandas/core/groupby/ops.py:106]: prob should be after the class variables / methods, but nbd

File: pandas/core/groupby/ops.py [hunk 3]
```diff
@@ -637,43 +681,23 @@ def _cython_operation(
...
-        if how == "rank":
-            out_dtype = "float"
-        else:
-            if is_numeric:
-                out_dtype = f"{values.dtype.kind}{values.dtype.itemsize}"
-            else:
-                out_dtype = "object"
 
-        codes, _, _ = self.group_info
+        out_shape = cy_op.get_output_shape(ngroups, values)
```
Comment (jreback) [inline pandas/core/groupby/ops.py:690]: future, would just have a single method here where you pass `ngroup`, `values`, *may* be able to determine `is_numeric` inside the WrappedCythonOps (instead of here), and just return the items you need, though you can take it a step further and have a `WrappedAggregateOp` and `WrappedTransformOp`, something like

```
op = WrappedCythonOps(kind, how, values.....) # other args
# op is WrappedTransform / WrappedAggregate
result = op.get_result()
```

**Review by jreback — APPROVED** [review]
looks fine, some suggestions for possible followups

**Comment by jorisvandenbossche** [comment]
@jbrockmendel can you hold off a bit with such refactoring changes to groupby/ops.py until https://github.com/pandas-dev/pandas/pull/40672#issuecomment-811906908 is resolved?

**Comment by jbrockmendel** [comment]
> can you hold off a bit with such refactoring

Sure.  The two things I have queued up in this area are 1) a refactor aimed specifically at ameliorating the linked concern and 2) a bugfix where we currently cast int64s to float64s in a lossy way.  When I post them ill put them in "draft" mode and cc you on them.
